pkg/deploy/depoy.go

```diff
@@ -0,0 +1,247 @@
+package deploy
+
+// Copyright (c) Microsoft Corporation.
+// Licensed under the Apache License 2.0.
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"os"
+	"time"
+
+	azcompute "github.com/Azure/azure-sdk-for-go/services/compute/mgmt/2019-03-01/compute"
+	azresources "github.com/Azure/azure-sdk-for-go/services/resources/mgmt/2018-05-01/resources"
+	"github.com/Azure/go-autorest/autorest"
+	"github.com/Azure/go-autorest/autorest/to"
+	"github.com/sirupsen/logrus"
+
+	"github.com/Azure/ARO-RP/pkg/util/azureclient/mgmt/compute"
+	"github.com/Azure/ARO-RP/pkg/util/azureclient/mgmt/network"
+	"github.com/Azure/ARO-RP/pkg/util/azureclient/mgmt/resources"
+	"github.com/Azure/ARO-RP/pkg/util/ready"
+)
+
+var _ Deployer = (*deployer)(nil)
+
+type Deployer interface {
+	PreDeploy(context.Context, *logrus.Entry) (string, error)
+	Deploy(context.Context, *logrus.Entry, string) error
+	Upgrade(context.Context, *logrus.Entry) error
+}
+
+type deployer struct {
+	log         *logrus.Entry
+	deployments resources.DeploymentsClient
+	vmss        compute.VirtualMachineScaleSetsClient
+	vmssvms     compute.VirtualMachineScaleSetVMsClient
+	network     network.PublicIPAddressesClient
+
+	cli *http.Client
+
+	parameters        map[string]interface{}
+	version           string
+	resourceGroupName string
+	subscriptionID    string
+	location          string
+}
+
+// New initiates new deploy utility object
+func New(ctx context.Context, log *logrus.Entry, authorizer autorest.Authorizer, gitCommit string) (*deployer, error) {
+	d := &deployer{
+		log:               log,
+		version:           gitCommit,
+		resourceGroupName: os.Getenv("AZURE_RP_RESOURCEGROUP_NAME"),
+		subscriptionID:    os.Getenv("AZURE_SUBSCRIPTION_ID"),
+	}
+
+	d.cli = &http.Client{
+		Timeout: 5 * time.Second,
+	}
+
+	// extract location from rg
+	gCli := resources.NewGroupsClient(d.subscriptionID, authorizer)
+	group, err := gCli.Get(ctx, d.resourceGroupName)
+	if err != nil {
+		return nil, err
+	}
+	d.location = *group.Location
```

Comment:
follow up: both group name and location should go into the input manifest

Reply:
ack